source/common/http/async_client_impl.cc

```diff
@@ -105,8 +105,14 @@ void AsyncStreamImpl::sendHeaders(HeaderMap& headers, bool end_stream) {
 }
 
 void AsyncStreamImpl::sendData(Buffer::Instance& data, bool end_stream) {
-  router_.decodeData(data, end_stream);
-  decoding_buffer_ = &data;
+  FilterDataStatus status = router_.decodeData(data, end_stream);
+
+  if (status == FilterDataStatus::StopIterationAndBuffer) {
+    decoding_buffer_ = &data;
```

Comment:
This is still not right. To probably support buffering you actually need to buffer the data. For example, have Buffer::OwnedImpl, and add each data frame to it. I don't think this is worth supporting, as I don't think there is any reasonable case where for a streaming request we will need buffering. I would refactor the code slightly in such a way where in the base streaming case, you don't support buffering (throw an exception if something results in buffering being needed), but in the non-streaming case you do what we did previously, which is to just return the request's body data.